File: LetterConbinationOfAPhoneNumber.py

D = {
    "1": None,
    "2": "abc",
    "3": "def",
    "4": "ghi",
    "5": "jkl",
    "6": "mno",
    "7": "pqrs",
    "8": "tuv",
    "9": "wxyz",
    "0": " ",
}


class Solution:
    ans = []
    digits = None

    def dfs(self, buff, idx):
        if idx >= len(self.digits):
            self.ans.append(buff)
            return
        for c in D[self.digits[idx]]:
            self.dfs(buff + c, idx + 1)


    # @return a list of strings
    def letterCombinations(self, digits):
        self.ans = []
        self.digits = digits
        self.dfs("", 0)
        return self.ans

File: LetterConbinationOfAPhoneNumber.cpp
#include <vector>
#include <string>
#include <iostream>

using std::vector;
using std::string;
using std::cout;
using std::endl;


#include <cstdlib>
using std::atoi;
struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL){}
};




#include <algorithm>
#include <cstring>

const vector<string> D = {" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

class Solution{
  public:
    
    vector<string> ans;
    string digits;
    typedef string::size_type sz;

    void dfs(string &buff, sz idx){
        if(idx >= digits.length()){
            ans.push_back(buff);
            return;
        }
        const string &s = D[digits[idx]-'0'];
        for(auto &c: s){
            buff += c;
            dfs(buff, idx + 1);
            buff.pop_back();
        }
    }

    vector<string> letterCombinations(string Digits){
        ans.clear();
        digits = std::move(Digits);
        string s = "";
        dfs(s, 0);
        return ans;
    }

};



int main(){
    Solution slu;
    using std::cin;

    auto ans = slu.letterCombinations("23");

    for(auto &s: ans){
        cout << s << endl;
    }

    return 0;
}


